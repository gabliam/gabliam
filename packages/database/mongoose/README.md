[![Build Status][build-image]][build-url]
[![NPM version][npm-image]][npm-url]

# mongoose
mongoose plugin for gabliam


# License

  MIT

[build-image]: https://img.shields.io/travis/gabliam/gabliam/master.svg?style=flat-square
[build-url]: https://travis-ci.org/gabliam/gabliam
[npm-image]: https://img.shields.io/npm/v/@gabliam/mongoose.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@gabliam/mongoose
