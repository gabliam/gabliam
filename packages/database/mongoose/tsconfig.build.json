{
  "extends": "../../../tsconfig.base",
  "compilerOptions": {
    "sourceMap": false,
    "outDir": "../../../dist/packages/database/mongoose",
    "baseUrl": ".",
    "rootDir": ".",
    "paths": {
      "@gabliam/*": [
        "../../../dist/packages/broker/*",
        "../../../dist/packages/cache/*",
        "../../../dist/packages/core/*",
        "../../../dist/packages/database/*",
        "../../../dist/packages/web/*",
        "../../../dist/packages/tools/*"
      ]
    }
  },
  "include":[
    "index.ts",
    "src/**/*.ts"
  ]
}
