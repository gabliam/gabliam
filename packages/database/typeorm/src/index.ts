import { TypeOrmPlugin } from './typeorm-plugin';

export { ConnectionManager } from './connection-manager';
export * from './connection-options-reader';
export * from './constant';
export * from './typeorm';
export { TypeOrmPlugin as default };
