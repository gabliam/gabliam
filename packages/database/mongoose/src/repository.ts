import { IRead, IWrite } from './interfaces';
import { mongoose } from './mongoose';

export class Repository<T>
  implements IRead<T & mongoose.Document>, IWrite<T, T & mongoose.Document> {
  public model: mongoose.Model<T & mongoose.Document>;

  constructor(schemaModel: mongoose.Model<T & mongoose.Document>) {
    this.model = schemaModel;
  }

  create(item: T): Promise<T & mongoose.Document> {
    return this.model.create(<any>item);
  }

  findAll(): Promise<(T & mongoose.Document)[]> {
    return this.model.find({}).exec();
  }

  update(_id: string | mongoose.Types.ObjectId, item: Object): Promise<any> {
    return this.model
      .updateOne(<any>{ _id: this.toObjectId(_id) }, <any>item)
      .exec();
  }

  delete(_id: string | mongoose.Types.ObjectId): Promise<void> {
    return (this.model
      .deleteOne(<any>{ _id: this.toObjectId(_id) })
      .exec() as unknown) as Promise<void>;
  }

  findById(_id: string): Promise<(T & mongoose.Document) | null> {
    return this.model.findById(_id).exec();
  }

  findOne(cond?: Object): mongoose.Query<(T & mongoose.Document) | null> {
    return this.model.findOne(<any>cond);
  }

  find(
    conditions: Object,
    projection?: Object,
    options?: Object,
  ): mongoose.Query<(T & mongoose.Document)[]> {
    return this.model.find(<any>conditions, projection!, options!);
  }

  deleteAll() {
    return this.model.deleteMany({}).exec();
  }

  private toObjectId(
    _id: string | mongoose.Types.ObjectId,
  ): mongoose.Types.ObjectId {
    if (mongoose.Types.ObjectId.isValid(_id)) {
      return <mongoose.Types.ObjectId>_id;
    }

    /* istanbul ignore next */
    return mongoose.Types.ObjectId.createFromHexString(<string>_id);
  }
}
