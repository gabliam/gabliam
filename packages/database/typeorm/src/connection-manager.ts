import { createConnections } from 'typeorm';
import { GabliamConnectionOptionsReader } from './connection-options-reader';
import { TypeormConnectionNotFoundError } from './errors';
import { Connection } from './typeorm';

export class ConnectionManager {
  private connections!: Connection[];

  constructor(public connectionOptionsReader: GabliamConnectionOptionsReader) {}

  async open() {
    this.connections = await createConnections(
      await this.connectionOptionsReader.all()
    );
  }

  async close() {
    await Promise.all(this.connections.map(c => c.close()));
  }

  getConnection(name: string) {
    const connection = this.connections.find(c => c.name === name);
    if (!connection) {
      throw new TypeormConnectionNotFoundError(name);
    }
    return connection;
  }

  getDefaultConnection() {
    const connection = this.connections.find(c => c.name === 'default');
    if (connection === undefined) {
      return this.connections[0];
    } else {
      return connection;
    }
  }
}
