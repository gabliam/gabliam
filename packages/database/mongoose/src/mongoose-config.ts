import {
  Bean,
  Container,
  Init,
  inject,
  InjectContainer,
  INJECT_CONTAINER_KEY,
  PluginConfig,
  Value,
} from '@gabliam/core';
import d from 'debug';
import { configurationValidator } from './configuration-validator';
import { MongooseConnectionManager } from './connection-manager';
import { LIST_DOCUMENT } from './constants';
import { MongoConfigIsMandatoryError } from './errors';
import { MongooseConfiguration } from './interfaces';
import { MongooseConnection } from './mongoose-connection';

const debug = d('Gabliam:Plugin:mongoose');

@InjectContainer()
@PluginConfig()
export class PluginMongooseConfig {
  @Value('application.mongoose', configurationValidator)
  mongooseConfiguration!: MongooseConfiguration | MongooseConfiguration[];

  listDocument: Function[];

  constructor(@inject(LIST_DOCUMENT) listDocument: Function[]) {
    debug('constructor PluginMongooseConfig');
    this.listDocument = listDocument;
  }

  // when all bean are created, we create bean Connection for back compat
  @Init()
  async init() {
    const container: Container = (<any>this)[INJECT_CONTAINER_KEY];
    const connectionManager = container.get(MongooseConnectionManager);
    await connectionManager.open();

    // for back compat
    container
      .bind(MongooseConnection)
      .toConstantValue(connectionManager.getDefaultConnection());
  }

  @Bean(MongooseConnectionManager)
  createManager() {
    debug('mongooseConfiguration', this.mongooseConfiguration);
    if (!this.mongooseConfiguration) {
      throw new MongoConfigIsMandatoryError();
    }
    let mongooseConfiguration: MongooseConfiguration[];
    if (Array.isArray(this.mongooseConfiguration)) {
      mongooseConfiguration = this.mongooseConfiguration;
    } else {
      mongooseConfiguration = [this.mongooseConfiguration];
    }
    const listDocument = Array.isArray(this.listDocument)
      ? this.listDocument
      : [];
    return new MongooseConnectionManager(mongooseConfiguration, listDocument);
  }
}
