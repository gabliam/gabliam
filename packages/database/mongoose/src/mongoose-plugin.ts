import {
  Scan,
  Registry,
  Plugin,
  GabliamPlugin,
  Container,
} from '@gabliam/core';
import { TYPE, LIST_DOCUMENT } from './constants';
import d from 'debug';
const debug = d('Gabliam:Plugin:mongoose');
import { MongooseConnectionManager } from './connection-manager';

@Plugin()
@Scan()
export class MongoosePlugin implements GabliamPlugin {
  bind(container: Container, registry: Registry) {
    const documents = registry.get(TYPE.Document).map(({ id, target }) => {
      return target;
    });
    debug('list documents', documents);
    container.bind<any>(LIST_DOCUMENT).toConstantValue(documents);
  }

  async destroy(container: Container, registry: Registry) {
    const connectionManager = container.get(MongooseConnectionManager);
    await connectionManager.close();
  }
}
