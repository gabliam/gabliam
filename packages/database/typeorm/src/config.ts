import {
  Bean,
  Container,
  Init,
  inject,
  InjectContainer,
  INJECT_CONTAINER_KEY,
  optional,
  PluginConfig,
  Value,
  ValueExtractor,
  VALUE_EXTRACTOR,
} from '@gabliam/core';
import d from 'debug';
import { ConnectionManager } from './connection-manager';
import { GabliamConnectionOptionsReader } from './connection-options-reader';
import {
  ConnectionOptionsBeanId,
  ENTITIES_TYPEORM,
  MIGRATIONS_TYPEORM,
  SUBSCRIBERS_TYPEORM,
} from './constant';
import { Connection, ConnectionOptions } from './typeorm';

const debug = d('Gabliam:Plugin:Typeorm');

@InjectContainer()
@PluginConfig()
export class PluginTypeormConfig {
  @Value('application.typeorm.connectionOptions')
  connectionOptions: ConnectionOptions | ConnectionOptions[] | undefined;

  entities: {
    entities: Function[];
    migrations: Function[];
    subscribers: Function[];
  };

  migrations: Function[];

  constructor(
    @inject(ConnectionOptionsBeanId)
    @optional()
    connectionOptions: ConnectionOptions | ConnectionOptions[],
    @inject(ENTITIES_TYPEORM) entities: Function[],
    @inject(MIGRATIONS_TYPEORM) migrations: Function[],
    @inject(SUBSCRIBERS_TYPEORM) subscribers: Function[]
  ) {
    debug('constructor PluginTypeormConfig', connectionOptions);
    this.connectionOptions = connectionOptions;
    this.entities = { entities, migrations, subscribers };
  }

  // when all bean are created, we create bean Connection for back compat
  @Init()
  async init() {
    const container: Container = (<any>this)[INJECT_CONTAINER_KEY];
    const connectionManager = container.get(ConnectionManager);

    // for back compat
    container
      .bind(Connection)
      .toDynamicValue(() => connectionManager.getDefaultConnection());
  }

  @Bean(GabliamConnectionOptionsReader)
  createGabliamConnectionOptionsReader() {
    debug('connectionOptions', this.connectionOptions);

    let connectionOptions: ConnectionOptions[] | undefined = undefined;
    if (Array.isArray(this.connectionOptions)) {
      connectionOptions = this.connectionOptions;
    } else if (this.connectionOptions) {
      connectionOptions = [this.connectionOptions];
    }

    const container: Container = (<any>this)[INJECT_CONTAINER_KEY];
    const valueExtractor = container.get<ValueExtractor>(VALUE_EXTRACTOR);

    return new GabliamConnectionOptionsReader(
      connectionOptions,
      this.entities,
      valueExtractor
    );
  }

  @Bean(ConnectionManager)
  createManager() {
    return new ConnectionManager(this.createGabliamConnectionOptionsReader());
  }
}
