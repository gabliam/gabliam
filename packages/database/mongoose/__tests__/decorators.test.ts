import { reflection } from '@gabliam/core';
import mongoose from 'mongoose';
import { Document, MUnit } from '../src';

test(`@Document('Hero') decorator`, () => {
  @Document('Hero')
  class Hero {
    static getSchema() {
      return new mongoose.Schema({
        name: {
          type: String,
          required: true,
        },
      });
    }
  }

  expect(reflection.annotations(Hero)).toMatchSnapshot();
});

test(`@Document(DocumentOptions without collectionName)  decorator`, () => {
  @Document({
    name: 'Hero',
    schema: new mongoose.Schema({
      name: {
        type: String,
        required: true,
      },
    }),
  })
  class Hero {}

  expect(reflection.annotations(Hero)).toMatchSnapshot();
});

test(`@Document(DocumentOptions)  decorator`, () => {
  @Document({
    name: 'Hero',
    collectionName: 'HeroCollection',
    schema: new mongoose.Schema({
      name: {
        type: String,
        required: true,
      },
    }),
  })
  class Hero {}

  expect(reflection.annotations(Hero)).toMatchSnapshot();
});

describe('errors', () => {
  test('no getSchema', () => {
    expect(() => {
      @Document('Hero')
      class Hero {}
      // tslint:disable-next-line:no-unused-expression
      new Hero();
    }).toThrowError();
  });

  test('getSchema return bad info', () => {
    expect(() => {
      @Document('Hero')
      class Hero {
        static getSchema() {
          return 'lol';
        }
      }
      // tslint:disable-next-line:no-unused-expression
      new Hero();
    }).toThrowError();
  });
});

describe('MUnit decorators', () => {
  test('should add MUnit metadata to a class when decorating a method with @MUnit', () => {
    @MUnit('default')
    class TestEntity {}

    expect(reflection.annotations(TestEntity)).toMatchSnapshot();
  });

  test('should fail when decorated multiple times with @MUnit', () => {
    expect(function() {
      @MUnit('default')
      @MUnit('default2')
      class TestBean {}

      // tslint:disable-next-line:no-unused-expression
      new TestBean();
    }).toThrowError();
  });
});
