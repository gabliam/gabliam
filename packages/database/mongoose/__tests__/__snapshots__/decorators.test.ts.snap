// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Document('Hero') decorator 1`] = `
Array [
  DecoratorFactory {
    "name": "Hero",
    "schema": Schema {
      "$id": 1,
      "_indexes": Array [],
      "_userProvidedOptions": Object {},
      "aliases": Object {},
      "callQueue": Array [],
      "childSchemas": Array [],
      "inherits": Object {},
      "methodOptions": Object {},
      "methods": Object {},
      "nested": Object {},
      "obj": Object {
        "name": Object {
          "required": true,
          "type": [Function],
        },
      },
      "options": Object {
        "_id": true,
        "autoIndex": null,
        "bufferCommands": true,
        "capped": false,
        "discriminatorKey": "__t",
        "id": true,
        "minimize": true,
        "noId": false,
        "noVirtualId": false,
        "read": null,
        "shardKey": null,
        "strict": true,
        "typeKey": "type",
        "validateBeforeSave": true,
        "versionKey": "__v",
      },
      "paths": Object {
        "_id": ObjectId {
          "_index": null,
          "defaultValue": [Function],
          "getters": Array [],
          "instance": "ObjectID",
          "options": SchemaObjectIdOptions {
            "auto": true,
            "type": [Function],
          },
          "path": "_id",
          "setters": Array [
            [Function],
          ],
          "validators": Array [],
          Symbol(mongoose#schemaType): true,
        },
        "name": SchemaString {
          "_index": null,
          "enumValues": Array [],
          "getters": Array [],
          "instance": "String",
          "isRequired": true,
          "options": SchemaStringOptions {
            "required": true,
            "type": [Function],
          },
          "originalRequiredValue": true,
          "path": "name",
          "regExp": null,
          "requiredValidator": [Function],
          "setters": Array [],
          "validators": Array [
            Object {
              "message": "Path \`{PATH}\` is required.",
              "type": "required",
              "validator": [Function],
            },
          ],
          Symbol(mongoose#schemaType): true,
        },
      },
      "plugins": Array [],
      "query": Object {},
      "s": Object {
        "hooks": Kareem {
          "_posts": Map {},
          "_pres": Map {},
        },
      },
      "singleNestedPaths": Object {},
      "statics": Object {},
      "subpaths": Object {},
      "tree": Object {
        "_id": Object {
          "auto": true,
          "type": [Function],
        },
        "name": Object {
          "required": true,
          "type": [Function],
        },
      },
      "virtuals": Object {},
    },
  },
  DecoratorFactory {
    "autobind": false,
    "id": [Function],
    "type": "Document",
  },
]
`;

exports[`@Document(DocumentOptions without collectionName)  decorator 1`] = `
Array [
  DecoratorFactory {
    "name": "Hero",
    "schema": Schema {
      "$id": 2,
      "_indexes": Array [],
      "_userProvidedOptions": Object {},
      "aliases": Object {},
      "callQueue": Array [],
      "childSchemas": Array [],
      "inherits": Object {},
      "methodOptions": Object {},
      "methods": Object {},
      "nested": Object {},
      "obj": Object {
        "name": Object {
          "required": true,
          "type": [Function],
        },
      },
      "options": Object {
        "_id": true,
        "autoIndex": null,
        "bufferCommands": true,
        "capped": false,
        "discriminatorKey": "__t",
        "id": true,
        "minimize": true,
        "noId": false,
        "noVirtualId": false,
        "read": null,
        "shardKey": null,
        "strict": true,
        "typeKey": "type",
        "validateBeforeSave": true,
        "versionKey": "__v",
      },
      "paths": Object {
        "_id": ObjectId {
          "_index": null,
          "defaultValue": [Function],
          "getters": Array [],
          "instance": "ObjectID",
          "options": SchemaObjectIdOptions {
            "auto": true,
            "type": [Function],
          },
          "path": "_id",
          "setters": Array [
            [Function],
          ],
          "validators": Array [],
          Symbol(mongoose#schemaType): true,
        },
        "name": SchemaString {
          "_index": null,
          "enumValues": Array [],
          "getters": Array [],
          "instance": "String",
          "isRequired": true,
          "options": SchemaStringOptions {
            "required": true,
            "type": [Function],
          },
          "originalRequiredValue": true,
          "path": "name",
          "regExp": null,
          "requiredValidator": [Function],
          "setters": Array [],
          "validators": Array [
            Object {
              "message": "Path \`{PATH}\` is required.",
              "type": "required",
              "validator": [Function],
            },
          ],
          Symbol(mongoose#schemaType): true,
        },
      },
      "plugins": Array [],
      "query": Object {},
      "s": Object {
        "hooks": Kareem {
          "_posts": Map {},
          "_pres": Map {},
        },
      },
      "singleNestedPaths": Object {},
      "statics": Object {},
      "subpaths": Object {},
      "tree": Object {
        "_id": Object {
          "auto": true,
          "type": [Function],
        },
        "name": Object {
          "required": true,
          "type": [Function],
        },
      },
      "virtuals": Object {},
    },
  },
  DecoratorFactory {
    "autobind": false,
    "id": [Function],
    "type": "Document",
  },
]
`;

exports[`@Document(DocumentOptions)  decorator 1`] = `
Array [
  DecoratorFactory {
    "collectionName": "HeroCollection",
    "name": "Hero",
    "schema": Schema {
      "$id": 3,
      "_indexes": Array [],
      "_userProvidedOptions": Object {},
      "aliases": Object {},
      "callQueue": Array [],
      "childSchemas": Array [],
      "inherits": Object {},
      "methodOptions": Object {},
      "methods": Object {},
      "nested": Object {},
      "obj": Object {
        "name": Object {
          "required": true,
          "type": [Function],
        },
      },
      "options": Object {
        "_id": true,
        "autoIndex": null,
        "bufferCommands": true,
        "capped": false,
        "discriminatorKey": "__t",
        "id": true,
        "minimize": true,
        "noId": false,
        "noVirtualId": false,
        "read": null,
        "shardKey": null,
        "strict": true,
        "typeKey": "type",
        "validateBeforeSave": true,
        "versionKey": "__v",
      },
      "paths": Object {
        "_id": ObjectId {
          "_index": null,
          "defaultValue": [Function],
          "getters": Array [],
          "instance": "ObjectID",
          "options": SchemaObjectIdOptions {
            "auto": true,
            "type": [Function],
          },
          "path": "_id",
          "setters": Array [
            [Function],
          ],
          "validators": Array [],
          Symbol(mongoose#schemaType): true,
        },
        "name": SchemaString {
          "_index": null,
          "enumValues": Array [],
          "getters": Array [],
          "instance": "String",
          "isRequired": true,
          "options": SchemaStringOptions {
            "required": true,
            "type": [Function],
          },
          "originalRequiredValue": true,
          "path": "name",
          "regExp": null,
          "requiredValidator": [Function],
          "setters": Array [],
          "validators": Array [
            Object {
              "message": "Path \`{PATH}\` is required.",
              "type": "required",
              "validator": [Function],
            },
          ],
          Symbol(mongoose#schemaType): true,
        },
      },
      "plugins": Array [],
      "query": Object {},
      "s": Object {
        "hooks": Kareem {
          "_posts": Map {},
          "_pres": Map {},
        },
      },
      "singleNestedPaths": Object {},
      "statics": Object {},
      "subpaths": Object {},
      "tree": Object {
        "_id": Object {
          "auto": true,
          "type": [Function],
        },
        "name": Object {
          "required": true,
          "type": [Function],
        },
      },
      "virtuals": Object {},
    },
  },
  DecoratorFactory {
    "autobind": false,
    "id": [Function],
    "type": "Document",
  },
]
`;

exports[`MUnit decorators should add MUnit metadata to a class when decorating a method with @MUnit 1`] = `
Array [
  DecoratorFactory {
    "name": "default",
  },
]
`;
